system/test/tcptest.c

```diff
@@ -0,0 +1,224 @@
+
+#include <cgreen/cgreen.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+#include <poll.h>
+
+#include "tcp.h"
+
+#define IP_ADDR     "127.0.0.1"
+#define SERVER_PORT 42020
+#define CLIENT_PORT 42040
+
+int client_fd, server_fd, rc;
+
+// Name of test context
+Describe(TCPInterface);
+
+// Execute in the context immediately before each "Ensure" test
+BeforeEach(TCPInterface) {
+}
+
+// Execute after each test
+AfterEach(TCPInterface) {
+
+    // Close sockets
+    rc = TCPClose(client_fd);
+    assert_that(rc, is_equal_to(0));
+    rc = TCPClose(server_fd);
+    assert_that(rc, is_equal_to(0));
+
+}
+
+void setup_sockets(void) {
+
+    // Init both
+    client_fd = TCPClientInit();
+    server_fd = TCPServerInit(IP_ADDR, SERVER_PORT);
+    assert_that(client_fd, is_not_equal_to(-1));
+    assert_that(server_fd, is_not_equal_to(-1));
+
+    // Connect client to server
+    rc = TCPClientTryConnect(client_fd, IP_ADDR, SERVER_PORT);
+    assert_that(rc, is_not_equal_to(-1));
+
+    // Set server nonblocking for accept
+    rc = TCPSetNonBlocking(server_fd);
+    assert_that(rc, is_not_equal_to(-1));
+
+    // Accept client from server
+    rc = TCPServerTryAccept(server_fd);
+    assert_that(rc, is_not_equal_to(-1));
+    // Old FD already closed, save new one
+    server_fd = rc;
+
+    // Set client nonblocking for future calls to read
+    rc = TCPSetNonBlocking(client_fd);
+    assert_that(rc, is_not_equal_to(-1));
+
+}
+
+
+/**** Start test suite ****/
+
+Ensure(TCPInterface, successfully_initializes_loopback) {
+
+    client_fd = TCPClientInit();
+    server_fd = TCPServerInit(IP_ADDR, SERVER_PORT);
+
+    assert_that(client_fd, is_not_equal_to(-1));
+    assert_that(server_fd, is_not_equal_to(-1));
+
+}
+
+Ensure(TCPInterface, client_fails_with_async_connect) {
+
+    client_fd = TCPClientInit();
+    assert_that(client_fd, is_not_equal_to(-1));
+
+    // Attempt to connect to a nonexistent server
+    rc = TCPClientTryConnect(client_fd, IP_ADDR, SERVER_PORT);
+    assert_that(rc, is_equal_to(-1));
+    assert_that(errno, is_equal_to(ECONNREFUSED));
+    // perror("TCPClientTryConnect failed (expected)");
+
+    // Open to be closed
+    server_fd = TCPServerInit(IP_ADDR, SERVER_PORT);
+
+}
+
+Ensure(TCPInterface, server_fails_with_async_accept) {
+
+    server_fd = TCPServerInit(IP_ADDR, SERVER_PORT);
+    assert_that(server_fd, is_not_equal_to(-1));
+
+    // Set nonblocking for attempted accept
+    rc = TCPSetNonBlocking(server_fd);
+    assert_that(rc, is_not_equal_to(-1));
+
+    // Attempt to accept a nonexistent client
+    rc = TCPServerTryAccept(server_fd);
+    assert_that(rc, is_equal_to(-1));
+    assert_that(errno, is_equal_to(EAGAIN));
+    // perror("TCPServerTryAccept failed (expected)");
+
+    // Open to be closed
+    client_fd = TCPClientInit();
+
+}
+
+Ensure(TCPInterface, client_server_connect_loopback) {
+
+    setup_sockets();
```

Comment:
Should this unit test have an assertion?

Reply:
The function setup_sockets() has assertions built in, but I made it better and more clear in the next rev